### PR TITLE
locatedb: fix startup fail due to systemd path capabilities

### PR DESCRIPTION
**Pull request by florianjacob**
###### Motivation for this change
On fresh NixOS 17.09 installs, `update-locatedb.service` continuously fails with:
```
update-locatedb.service: Failed at step NAMESPACE spawning /nix/store/bad6dlkkpkbk6qjdw8qhfbr1710bl0f9-unit-script/bin/update-locatedb-start: No such file or directory
```
This is because of the `ReadWritePaths=/var/cache` namespace restriction, as on new installs, `/var/cache` does not exist and therefore results in the `no such file or directory`. To reproduce this, delete your `/var/cache` directory.

I try to fix this by pulling out the mkdir part in a separate service. Note that this still fails on the *first* invocation with the same error, but creates the directory and then works on all subsequent invocations. I don't know why that happens. Maybe because the namespace restrictions are evaluated instantly when the unit is started, or because I did something wrong with the depency specification?

In any case, working the second day after being enabled is an improvement over always failing until the user creates `/var/cache` manually.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/misc/locate.nix [hunk 1]
```diff
@@ -125,13 +125,23 @@ in {
...
+
+    # the mkdir needs to be separated from main service
+    # because ReadWritePaths fails when the directory doesn't already exist
+    systemd.services.locatedb-setup =
+      { description = "Create locatedb state on first run";
+        script = ''
+            mkdir -m 0755 -p ${dirOf cfg.output}
+        '';
+        requiredBy = [ "update-locatedb.service" ];
+        before = [ "update-locatedb.service" ];
```
Comment (Mic92) [inline nixos/modules/misc/locate.nix]: I would probably make this of type `oneshot` and `RemainAfterExit=true`.

File: nixos/modules/misc/locate.nix [hunk 2]
```diff
@@ -125,13 +125,23 @@ in {
...
             ++ optional (isFindutils && cfg.pruneBindMounts) "findutils locate does not support skipping bind mounts";
-  
+
+    # the mkdir needs to be separated from main service
+    # because ReadWritePaths fails when the directory doesn't already exist
+    systemd.services.locatedb-setup =
+      { description = "Create locatedb state on first run";
+        script = ''
+            mkdir -m 0755 -p ${dirOf cfg.output}
+        '';
```
Comment (Mic92) [inline nixos/modules/misc/locate.nix]: The following should save building a dedicated script derivation:

```
serviceConfig.ExecStart = "${pkgs.coreutils}/bin/mkdir -m 0755 -p ${dirOf cfg.output}"
```

File: nixos/modules/misc/locate.nix [hunk 3]
```diff
@@ -147,8 +157,9 @@ in {
         serviceConfig.PrivateTmp = "yes";
         serviceConfig.PrivateNetwork = "yes";
         serviceConfig.NoNewPrivileges = "yes";
-        serviceConfig.ReadOnlyDirectories = "/";
-        serviceConfig.ReadWriteDirectories = dirOf cfg.output;
+        serviceConfig.ReadOnlyPaths = "/";
+        # TODO: one could try to further reduce this to cfg.output by pre-creating the file in locatedb-setup
+        serviceConfig.ReadWritePaths = dirOf cfg.output;
```
Comment (Mic92) [inline nixos/modules/misc/locate.nix]: `ConditionPathExists=` might also lead to better error messages.

**Comment by Mic92** [comment]
I wonder, if we just could use `systemd-tempfiles` instead of creating a dedicated service.

File: nixos/modules/misc/locate.nix (hunk 3 above)
Comment (florianjacob) [reply nixos/modules/misc/locate.nix]: Just tested it, sadly neither `ConditionPathExists` nor `AssertPathExists` seem to have any effect, it's always just the `NAMESPACE` error. 😞

**Comment by florianjacob** [comment]
Forgot about `systemd-tmpfiles`, turned out to be much simpler, I could remove the extra unit again and the first invocation problem vanished as well, of course.

@Mic92 thanks alot for the hint, just updated the PR.

**Comment by Mic92** [comment]
also in 17.09 now.

**Pull request merged**